### ISSUE TITLE
UI improvements

### ISSUE THREAD

**Issue by certik** [inducer/pudb#181]
I like PuDB since it runs in a terminal itself. I noticed that one of the inspirations was a DOS-based Turbo Pascal, which I used to use a lot. Its text user interface (TUI), called Turbo Vision, was excellent. Compared to it, all TUI that we have on linux feel unpolished, hard to navigate, bad mouse support etc. To understand what I mean, compare the look of the old Turbo Vision (you can find more screenshots e.g. of Turbo Pascal IDE itself [here](http://yvz.manisateknikservis.com/kutuphane/free-pascal/user/node7.htm)):
![](http://thomasjensen.com/software/quadra/quadra.gif)
or Turbo Debugger:
![](http://keyj.emphy.de/files/td.gif)
with PuDB:
![](https://camo.githubusercontent.com/a55e51e1bc45a43dde7de1c3b2ea2c0cb717f4fb/687474703a2f2f74696b65722e6e65742f7075622f707564622d73637265656e73686f742e706e67)
The PuDB just feels like a poor cousin. It's very unintuitive for me to see which of the subwindows has focus, it doesn't seem to react on mouse properly, and overall the look and feel seems very unpolished. I have no doubts that the pleasure of debugging can be vastly improved if the TUI itself is inviting.

I spent some time investigating what can be done about this. One question is philosophical --- does it makes sense to spend time doing text UI (TUI), when we can just use some rich GUI like Qt? For me the answer is definitely yes, as when using a terminal only, then my development environment looks exactly the same locally on a desktop, as well as remotely over ssh. And a modern terminal is pretty pleasurable to work with.

Another question is --- why not to move his into a browser? Similar to the [Jupyter notebook](http://jupyter.org/). I.e. you run the application locally or remotely in a terminal, forward the HTTP and use a local browser. Then you can use all kinds of great GUIs in the browser itself. That might be the way to go. I don't know.

Another similar idea is to improve the terminal itself, to be better at UI. E.g. there is the [Qt Console](http://jupyter.org/qtconsole/stable/), then it looks like a terminal, but can do images. It could in principle be extended to allow windows, menus, scrollbars, etc. Other, discontinued, projects are [TermKit](https://github.com/unconed/TermKit), [finalterm](https://github.com/p-e-w/finalterm) that tried to improve the terminal itself. Then there are projects to improve the shell experience ([Xiki](https://www.kickstarter.com/projects/xiki/xiki-the-command-revolution/description), [Xonsh](http://xon.sh)). Some people say that if `TermKit` was build using the [Electron](http://electron.atom.io/) platform today, it would be successful (see e.g. the screenshots of [Mancy](https://github.com/princejwesley/Mancy), which is build using `Electron`).

Some of these new tools might become the way to go. But in the meantime, I still see value in simply using the plain old terminal, and do the best one can do using it, because terminals are versatile, cross platform and ubiquitous.

There is an open source implementation of Turbo Vision, called [Free Vision](http://wiki.freepascal.org/Free_Vision) which is actively maintained, and the Free Pascal Compiler IDE is written in it, here is a screenshot:
![](https://upload.wikimedia.org/wikipedia/commons/3/34/FPIDE_1.0.10_de.png)
You can download it [here](http://www.freepascal.org/download.var), it works in a terminal (I tried it on Ubuntu 14.04), and it feels pretty much exactly like the old Turbo Pascal.

I think one would need to write a library (either improving [urwid](http://urwid.org/), or start a new one), that has such a nice TUI. I think it would be widely successful, since urwid itself is successful, and this would make it look nice and polished. There is a library for NodeJS called [Blessed](https://github.com/chjj/blessed) (see the screencast there), which I tried as well, and they have some nice looking things, but overall, it still feels less polished than Turbo Vision (or Free Vision). It's obviously a lot of work, but I think it is worth it, to have a modern, nice looking and polished TUI library.

A separate and relevant question is mouse support. For some reason most apps do not use mouse, it seems pudb doesn't use it either. Free Vision as well as Blessed work with mouse. I looked at it, and one has to use the [1003 mouse reporting mode](http://stackoverflow.com/questions/5966903/how-to-get-mousemove-and-mouseclick-in-bash/5970472#5970472) in a terminal, which makes the terminal app send ANSI escape sequences for mouse movement, that you just read in your code. For some reason, ncurses doesn't seem to support it (it can read mouse clicks, but not mouse movement --- which is useless, if you want to implement, say, a scrollbar). So I wrote my own [simple app](https://gist.github.com/certik/47910b9c9964ca55743f413a670240e6) that moved a character on a screen based on the mouse movement, here is a screencast:
[![asciicast](https://asciinema.org/a/612ignepn8g1ajv85fpqti33o.png)](https://asciinema.org/a/612ignepn8g1ajv85fpqti33o)
it also reads the mouse clicks (the top left character, that would have to be decoded). I tested in `gnome-terminal` in Ubuntu 14.04 and 16.04, as well as the default terminal on OS X El Capitan. I also tested over ssh (two hops in fact...) and in `screen`. It works like a charm everywhere, with default settings. So I think that terminal apps should support mouse by default (in addition to keyboard, not instead of it). In particular, this would be very useful for a scrollbar in the PuDB window, so that you can use it exactly like a regular scrollbar, say in a web browser, only this would be in the terminal itself. Minor point: note that my app shows the pointer in the text mode, but since the terminal (both on Linux and Mac) also shows the mouse pointer anyway, perhaps it's not necessary to show the "text mouse pointer" (like Turbo Pascal did). I noticed that both `Blessed` as well as `Turbo Pascal` for some reason have a bit different behavior of the mouse than one is used to (e.g. neither allows you to "grab" the scrollbar), which is a shame, it should behave just like on a desktop.

Conclusion: I would really like to see a modern looking TUI, with first class mouse support, and nice, polished interface, inspired by the style of Turbo Vision. And PuDB can then use it.

@inducer, let me know what your thoughts are about this.

**Comment by asmeurer** [comment]
I agree that PuDB should have mouse tracking support. There was an old PR that started to implement it (https://github.com/inducer/pudb/pull/15). It will probably need to be redone, since a lot of PuDB has changed since then. But it looks like Urwid does have mouse support, so it probably won't be hard (just add a lot of `listen_mouse_event` handlers). If you want to revive that PR, I'll help review it and get it merged (but I don't have time to work on it directly at the moment). 

@inducer said that he doesn't like mouse support because it breaks his ability to select things in the shell. The terminal emulator I use (iTerm2) lets you press a key (option) and it disables mouse tracking. I haven't used Linux in a while but I would expect any decent terminal emulator to have a similar feature. We could also have an option to disable it. I think it should be on by default, but we can debate that. 

The nice thing about Urwid vs. other libraries is that it's a GUI framework, that is, it works at a GUI abstraction level, with classes for things like dialogs and checkboxes and so on.  If you want to rewrite Urwid, well, good luck. That sounds like a lot of work. Maybe something like [prompt-toolkit](https://github.com/jonathanslenders/python-prompt-toolkit) can evolve in this direction, though. 

Urwid does have a browser backend, although I don't know how well it works, or how difficult it would be to port PuDB to it. I personally like that it's in the terminal. Like you said, I am already in the terminal all the time (I run IPython and emacs both in the terminal, so it fits nicely with that workflow). 

PuDB in general takes on a bit of the aesthetic that's common to terminal-based applications, including things like vim and emacs, which is that it requires a bit of knowledge of the right keys to use (with a bit of on-screen help), but once you know those keys, it's very easy, and faster than any GUI would be. With that being said, suggestions to improve the UX are welcome. Moving between panes is definitely one of the most [common issues](https://github.com/inducer/pudb#faq) people have. A menu bar is an interesting idea (does Urwid support menus?). I think @inducer has said that he doesn't want to waste any more vertical space, but I guess a menu bar could replace the current top bar.

**Comment by asmeurer** [comment]
Regarding scrollbars, those ought to be implemented in Urwid itself. A quick Google suggests that it doesn't have them.

**Comment by certik** [comment]
That's a good point about a full blown GUI (or TUI...) like Turbo Vision, as opposed to apps like Vim, Emacs, Mutt, where you need to learn the right keys (specific to the app), but once you do, it's very efficient. It's an open question whether you can have an intuitive GUI like Turbo Vision, but still allow to bind keys, so that you can have the efficient workflow.

However, it does seem that Turbo Vision style would waste some vertical space (e.g. the window border), as well as some horizontal space. My terminals are actually 80 characters wide (quite narrow, but I can have many terminals open at the same time, side by side, and I use 80 column limit in Python, C++, etc. anyway) and e.g. 85 rows tall (on my laptop it's more like 35 rows, since my screen is smaller). So at least in my case, I do seem to have plenty of vertical space.

Regarding rewriting urwid, that's a lot of work, and I don't have the time to do it myself. I wanted to mainly make sure in this issue, if this is something we even want.

**Comment by inducer** [comment]
@asmeurer Now that I know that Shift+mouse will give me my mouse selection back, I'm happy to enable mouse by default. Happy to consider a revived version of #15 (if it even needs reviving--let me know).

I personally like Urwid. It feels like it hits the right level of abstraction, and it gets a lot of tricky stuff right (unicode, double-wide characters, dealing with yucky, half-broken terminal emulators). It's lacking some things (scroll bars!), but I feel as though those as well as other missing pieces could fairly easily be added. Frames and such are nice to have, but they do come at the price of vertical space, which I find is at a premium even on this 4K monstrosity sitting in front of me. :)

tl;dr: TUI improvements: yes please, but I do think this can (and should!) be done incrementally. I won't have much time to help other than by reviewing patches. And I'd certainly appreciate smaller increments to keep the review effort manageable.

**Comment by asmeurer** [comment]
@certik I would ask @jonathanslenders, the prompt-toolkit guy if it's something he's interested in. That's currently the most actively developed Python terminal library from what I know, and it's showing up in a lot of places (e.g., xonsh uses it, and the next version of IPython will use it).

**Comment by jonathanslenders** [comment]
About prompt-toolkit and its direction:

In theory, it should definitely be possible to implement a UI like Pudb with prompt-toolkit. ptpdb is probably coming the closest (ptpdb is far from stable, btw.). However, keep in mind that until now, the main use case was to develop REPLs, so that's what is documented and tested best. Until now, there are only two full screen applications: pyvim and pymux. Both are very performant, but don't have advanced user controls like a UI toolkit typically provides.

So, I think that prompt-toolkit is probably lacking the user controls that Urwid offers. But in theory, all of that can be implemented on top of what we have right now, without changing anything in prompt-toolkit itself. However, it won't be easy without looking into the source code of prompt-toolkit. This isn't well documented yet. I will try to provide more examples/documentation for this in the coming months.

Mouse dragging support has also not yet been implemented. Maybe in the future, but it's not a priority. Click events (mouse up/down/scrolling) are handled by the framework.

Jonathan
